README.md
# User API
This is a simple REST API for users

# About the API

| ENDPOINT | HTTP VERB | ACTION |
|----------|-----------|--------|
|/users/   | GET       | get all users |
|/user/ID | GET       | get a single user by ID |
|/user/ID | PATCH    | update a user's info |
|/user/ID | DELETE  | delete a user  |
|/users/ |  POST   | create a new user |

### GET all users

```
[
 {"username":"user1",
   "email": "user1@test.com",
	 "password":"password"
 },
  {"username":"user2",
   "email": "user2@test.com",
     "password":"password"
   },  
]

```

That returns a list of resources


### GET a user by id

```
{"success":true,
  { "id":1
    "username":"user",
    "email":"user@test.com",
     "password":"password"
  }

}

```

This returns a user with an ID

### DELETE a user by ID

```
{
    "message": "User Deleted",
    "success": true,
    "user": {
        "email": "user@test.com",
        "id": 2,
        "password": "password",
        "username": "user"
    }
}




`i``


history.txt
  186  heroku run pip install pillow -a covidxray
  187  heroku logs --tail -a covidxray
  188  cd coding/python/covid-19-xray-image-classifier/
  189  heroku logs --tail -a covidxray
  190  export FLASK_APP=wsgi.py 
  191  export FLASK_DEBUG=1
  192  flask run
  193  clear
  194  flask run
  195  clear
  196  flask run
  197  cd helloworld/
  198  git status 
  199  git checkout v1.0.1
  200  git checkout -b v1.0.1
  201  npm install vuetify --save
  202  code .
  203  npm
  204  js
  205  sudo apt install nodejs
  206  sudo apt install npm
  207  sudo apt install npm --fix-missing
  208  sudo apt update
  209  sudo vim /etc/apt/sources.list
  210  sudo apt update
  211  sudo apt install npm --fix-missing
  212  sudo apt --fix-broken install
  213  node
  214  js
  215  npm install -g @vue/cli
  216  sudo npm install -g @vue/cli
  217  vue
  218  vue ui
  219  ls
  220  rm -rf helloworld/
  221  vue create helloworld
  222  cd helloworld/
  223  npm serve
  224  npm run serve
  225  rm -rf helloworld/
  226  vue create helloworld
  227  npm cache clean --force 
  228  vue create helloworld
  229  rm -rf helloworld/
  230  cd 
  231  rm -rf helloworld/
  232  vue create helloworld
  233  rm -rf helloworld/
  234  vue create helloworld
  235  sudo npm update -g @vue/cli
  236  vue create helloworld
  237  clear
  238  vue create helloworld
  239  sudo vue upgrade
  240  npm cache clean --force
  241  sudo npm cache clean --force
  242  sudo npm cache clean
  243  sudo npm cache clean --force
  244  sudo npm install -g @vue/cli
  245  vue create helloworld
  246  cd helloworld/
  247  npm run serve
  248  /bin/python3.9
  249  /usr/bin/python3 /home/jod35/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/pyvsc-run-isolated.py pip install -U black --user
  250  clear
  251  python3 app.py 
  252  git init
  253  touch .gitignore
  254  git add --all
  255  git commit -m "initial commit"
  256  git remote add origin https://github.com/jod35/Uploadr.git
  257  git push -u origin master
  258  clear
  259  python3 app.py 
  260  git add --all
  261  git commit "retrieve files from storage"
  262  git commit -m "retrieve files from storage"
  263  git push
  264  clear
  265  flask run
  266  git add --all
  267  git commit -m "successful file uploads"
  268  git push
  269  source /home/jod35/anaconda3/bin/activate
  270  conda activate base
  271  /home/jod35/anaconda3/bin/python /home/jod35/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/pyvsc-run-isolated.py pip install -U pylint
  272  git init
  273  git add --all
  274  git commit -m "initial commit"
  275  python3 app
  276  clear
  277  python3 app.py 
  278  git add --all
  279  git commit -m "successful file upload"
  280  git add --all
  281  git commit -m "edit .gitignore"
  282  git add --all
  283  git remote add origin https://github.com/jod35/Dropzone-flask.git
  284  git push -u origin master
  285  source /home/jod35/anaconda3/bin/activate
  286  conda activate base
  287  python3 app.py 
  288  pip3 install flask-dropzone
  289  mkdir dropzone-demo
  290  mv dropzone-demo/ coding/python/
  291  cd coding/python/dropzone-demo/
  292  code .
  293  cd coding/python/
  294  git clone https://github.com/jod35
  295  git clone https://github.com/jod35/anya.git
  296  rm -rf anya/
  297  git clone https://github.com/jod35/anya.git
  298  cd coding/python/
  299  ls
  300  cd ..
  301  ls
  302  cd 
  303  ls
  304  locate anya
  305  git clona anya
  306  git clona httpS://github.com/jod35/anya
  307  git clone httpS://github.com/jod35/anya
  308  git clone https://github.com/jod35/anya
  309  cd anya/
  310  code .
  311  git status
  312  git add --all
  313  git commit -m "fix :styles in paginate div"
  314  git push
  315  git add --all
  316  git commit -m "fix :zoom in on card images"
  317  git push
  318  zsh
  319  python3 manage.py runserver
  320  deconda
  321  python3 manage.py runserver
  322  pip3 install python-decouple whitenoise python-dotenv
  323  python3 manage.py runserver
  324  pip3 install dj_database_uri
  325  pip3 install dj_database_url
  326  python3 manage.py runserver
  327  pip3 install django_crispy_forms
  328  python3 manage.py runserver
  329  cd anya/
  330  git add --all
  331  git commit -m "fix :change page for password reset"
  332  git push
  333  git add --all
  334  git commit -m "fix :change images for homepage"
  335  git push
  336  clear
  337  sudo apt install plank
  338  cd coding/python/covid-19-xray-image-classifier/
  339  ls
  340  export FLASK_APP=wsgi.py 
  341  export FLASK_DEBUG=1
  342  flask run
  343  pip3 -U install flask
  344  pip3 install -U flask
  345  clear
  346  flask run
  347  lsof -i :5000
  348  sudo killall
  349  sudo kill 3242
  350  flask run
  351  clear
  352  flask run
  353  python3 wsgi.py 
  354  clear
  355  python3 wsgi.py 
  356  sudo apt install gnome
  357  clear
  358  python3 wsgi.py 
  359  plank 
  360  time
  361  date
  362  clear
  363  cal
  364  python3 wsgi.py 
  365  clear
  366  python3 wsgi.py 
  367  git add --all
  368  git commit -m "successful file upload"
  369  git push
  370  python3 wsgi.py 
  371  git add --all
  372  git commit -m "adding some styling"
  373  git oush
  374  git push
  375  sudo apt install golang
  376  sudo apt install golang --fix-missing
  377  sudo apt autoremove
  378  clear
  379  bash -c "$(wget -q -O - https://linux.kite.com/dls/linux/current)"
  380  faraday-manage --help
  381  service 
  382  clear
  383  export FLASK_APP=wsgi.py 
  384  export FLASK_DEBUG=1
  385  flask run
  386  git add --all
  387  git commit -m "successful return of data from server to broswer"
  388  git push
  389  git add --all
  390  git commit -m "successful return of data from server to broswer"
  391  git add --all
  392  git commit -m "successful return of data from server to broswer"
  393  clear
  394  export FLASK_APP=wsgi.py 
  395  export FLASK_DEBUG=1
  396  flask run

api.py
from flask import (Flask,
                    request,
                    jsonify,
                    make_response)

from marshmallow import Schema,fields
from flask_sqlalchemy import SQLAlchemy
from werkzeug.security import generate_password_hash,check_password_hash
import os


BASE_DIR=os.path.realpath(os.path.dirname(__file__))



app=Flask(__name__)

app.config["SQLALCHEMY_DATABASE_URI"]='sqlite:///'+ os.path.join(BASE_DIR,'myapp.db')
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"]=False
app.config["SQLALCHEMY_ECHO"]=True


db=SQLAlchemy(app)


class User(db.Model):
    id=db.Column(db.Integer(),primary_key=True)
    username=db.Column(db.String(255),nullable=False)
    email=db.Column(db.String(80),nullable=False)
    password=db.Column(db.Text)


    """def __init__(self,username,email,password):
        self.username=username
        self.email=email
        self.password=password
    """


    def save(self):
        db.session.add(self)
        db.session.commit()

    def delete(self):
        db.session.delete(self)
        db.session.commit()


    def create_password(self,password):
        self.password=generate_password_hash(password)

    def check_password(self,password):
        return check_password_hash(self.password,password)


class userSchema(Schema):
    id=fields.Integer()
    username=fields.Str()
    email=fields.Str()
    password=fields.Str()



@app.route('/users',methods=['GET'])
def get_users():
    users=User.query.order_by(User.id.desc()).all()

    users_=userSchema(many=True).dump(users)


    return make_response(jsonify({
        "users":users_}),200)

@app.route('/users',methods=['POST'])
def create_user():
    data=request.get_json()
    username=data.get('username')
    email=data.get('email')
    password=data.get('password')

    new_user=User(username=username,email=email)

    new_user.create_password(password)

    new_user.save()

    user_=userSchema().dump(new_user)


    return make_response(jsonify({"user":user_,"message":"User Created"}),201)

@app.route('/user/<int:id>')
def get_user_by_id(id):
    user=User.query.get_or_404(id)

    user_=userSchema().dump(user)


    return make_response(jsonify({"success":True,"user":user_}))

@app.route('/user/<int:id>',methods=['PATCH'])
def update_user_infor(id):
    user=User.query.get_or_404(id)

    data=request.get_json()


    user.username=data.get('username')
    user.email=data.get('email')


    db.session.commit()
    

    user_=userSchema().dump(user)


    return make_response(jsonify({"message":"Info Updated","success":True,"user":user_}))


@app.route('/user/<int:id>',methods=['DELETE'])
def delete_user(id):
    user=User.query.get_or_404(id)

    user.delete()

    user_=userSchema().dump(user)

    return make_response(jsonify({"message":"User Deleted","success":True,"user":user_}))


@app.errorhandler(404)
def not_found(err):
    return make_response(jsonify({"message":"Not Found"}))

@app.errorhandler(500)
def server_error(err):
    return make_response(jsonify({"message":"Something went wrong"}))





@app.shell_context_processor
def make_shell_context():
    return {
            "app":app,
            "db":db,
            "User":User}



if __name__ == "__main__":
    app.run(debug=True)
